### PR TITLE
[WIP] Feature: (iOS) Manual record and capture to prevent black flicker

### PR DESCRIPTION
**Pull request by reime005**
I recoded the part that causes the iOS black flicker when starting a record. This also brings the possibility of for example face recognition and recording at the same time. 

This should fix issues like #1249. 

This has to be tested and reviewed, since I refactored and changed a lot of code.

Related: #1645, #1542.

Any input is highly appreciated!

### CODE REVIEW

File: ios/RN/RNAssetWriter.m [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    AVAssetWriter* assetWriter;
+    AVAssetWriterInput* assetVideoWriterInput;
+    AVAssetWriterInput* assetAudioWriterInput;
+    AVAssetWriterInputPixelBufferAdaptor* pixelBufferAdaptor;
+    
+    dispatch_queue_t writerQueue;
+}
+
+- (BOOL)initRecording {
+    writerQueue = dispatch_queue_create("com.mariusreimer.writerQueue", dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_USER_INTERACTIVE, 0));
```
Comment (listicos) [inline ios/RN/RNAssetWriter.m]: Hello, I noticed a couple of things that are not good for an open source project, like this label.

File: ios/RN/RNAssetWriter.m [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+}
+
+- (void)initAssetWriterInputs {
+    // Video writer
+    assetVideoWriterInput = [AVAssetWriterInput
+                             assetWriterInputWithMediaType:AVMediaTypeVideo
+                             outputSettings:_videoSettings];
+    assetVideoWriterInput.expectsMediaDataInRealTime = YES;
+    
+//    CGAffineTransform transform = assetVideoWriterInput.transform;
```
Comment (listicos) [inline ios/RN/RNAssetWriter.m]: Do we need this ?

File: ios/RN/RNCameraManager.m [hunk 3]
```diff
@@ -203,11 +204,31 @@ + (NSDictionary *)faceDetectorConstants
...
+    }
+}
 
 RCT_CUSTOM_VIEW_PROPERTY(faceDetectorEnabled, BOOL, RNCamera)
 {
-    view.isDetectingFaces = [RCTConvert BOOL:json];
-    [view updateFaceDetecting:json];
+    BOOL newFaceDetectorEnabled = [RCTConvert BOOL:json];
+    view.isDetectingFaces = newFaceDetectorEnabled;
+    [view updateFaceDetecting:newFaceDetectorEnabled];
```
Comment (listicos) [inline ios/RN/RNCameraManager.m]: This is causing an error.

File: ios/RN/RNAssetWriter.m [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        if (CMTIME_IS_INVALID(firstVideoTimestamp) &&
+            !CMTIME_IS_INVALID(firstAudioTimeStamp)) {
+            firstVideoTimestamp = frameTime;
+            
+            [assetWriter startWriting];
+            
+            CMTime startTime = CMTimeAdd(firstAudioTimeStamp, firstVideoTimestamp);
+            [assetWriter startSessionAtSourceTime:startTime];
+            
+            RCTLogWarn(@"BEGIN writing at %f", CMTimeGetSeconds(startTime));
```
Comment (listicos) [inline ios/RN/RNAssetWriter.m]: Do we need a warning here?

File: ios/RN/RNAssetWriter.m [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+//TODO [reime005] add callback for when really finished
+- (void)finishWritingWithCompletionHandler:(void (^)(void))handler NS_AVAILABLE(10_9, 6_0) {
+    if (assetWriter == nil) {
+        RCTLogWarn(@"[RNAssetWriter] - not initialized!");
+        return;
+    }
+    
+    bWriting = NO;
+    RCTLogWarn(@"finished. fps: %.1f", frameNumber/_maxDuration);
```
Comment (listicos) [inline ios/RN/RNAssetWriter.m]: Do we need this warning ?

**Comment by listicos** [comment]
Also the PR has a conflicted file.

File: ios/FaceDetector/RNFaceDetectorManager.h [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
 - (instancetype)initWithSessionQueue:(dispatch_queue_t)sessionQueue delegate:(id <RNFaceDetectorDelegate>)delegate;
 
-- (void)setIsEnabled:(id)json;
+- (void)setIsEnabled:(BOOL)newFaceDetecting;
```
Comment (sibelius) [inline ios/FaceDetector/RNFaceDetectorManager.h]: why this?

File: ios/RN/RNAssetWriter.m [hunk 7]
```diff
@@ -0,0 +1,330 @@
+//
+//  RNAssetWriter.m
+//  RNCamera
+//
+//  Created by Marius Reimer on 10/6/18.
+//
+
+#import "RNAssetWriter.h"
+
+@implementation RNAssetWriter
```
Comment (sibelius) [inline ios/RN/RNAssetWriter.m]: why we need this?

File: src/RNCamera.js [hunk 8]
```diff
@@ -431,10 +441,6 @@ export default class Camera extends React.Component<PropsType, StateType> {
       newProps.googleVisionBarcodeDetectorEnabled = true;
     }
 
-    if (props.onFacesDetected) {
```
Comment (sibelius) [inline src/RNCamera.js]: will this break?

**Comment by reime005** [comment]
thanks for the review. I forgot to remove some things as you mentioned, I will work on this. The 'black flicker' always happens when you change settings related to the camera. For example setting flash, orientation, preset (quality) or mirroring. This then needs to be moved to the constructor.

Apparently I didn't find a way (yet) to do face detection while recording with the google framework

**Comment by n1ru4l** [comment]
Closed in favor of #1886

**Pull request closed without merging**